src/core/primitives.js

```diff
@@ -279,6 +279,21 @@ class Ref {
     return `${this.num}R${this.gen}`;
   }
 
+  static fromString(str) {
+    let ref = RefCache[str];
+    if (ref) {
+      return ref;
+    }
+    const [num, gen] = str.split("R", 2);
+    if (isNaN(num) || (gen && isNaN(gen))) {
+      return null;
+    }
```

Comment:
Yep I knew that but I thought it was enough.
Do we have to be that much defensive ?
Aren't devs (other than us) supposed to pass the correct data ?

Reply:
Given that this is accessible through the API, and a third-party user may modify the contents of the `AnnotationStorage`, I think that we should ensure that we don't generate any bogus `Ref`s here. Note that even we may, at some future point, end up *accidentally* passing incorrect data this method.

Besides, considering how rarely this will be used I don't really expect proper validate to affect performance noticeably here.